File: MapXtreme/6.6/Samples/WebControlsSourceCode/cs/Design/MapControlDesigner.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.IO;
using System.Reflection;
using System.Web.UI.Design;
using System.Web.UI;
using System.Xml;
using EnvDTE;
using System.Runtime.InteropServices;


namespace MapInfo.WebControls
{
	/// <summary>
	/// Designer for MapControl.
	/// </summary>
	/// <remarks>This designer doesn't have lot of functionality, it just displays a brief  message about the usage. It also
	/// updates web.config file with information about MapXtreme session activator and httphandler.</remarks>
	public class MapControlDesigner : ControlDesigner
	{
		private MapControl _mapControl;

		///	<summary>Updates the web.config file with information about MapXtreme assemblies and handlers.</summary>
		///	<remarks>The information added to the web.config file is the httphandler, session activator settings and appsetting lines if they don't
		///	exist (these lines contain pooling, preloaded workspaces etc.).
		///	</remarks>
		/// <param name="documentURL">Url of the form</param>
		public static void UpdateForMapControl(string documentURL) {
			// Open the web.config file as xmldoc
			string webConfigPath;
			XmlDocument xmlDoc = WebDesignerUtility.GetWebConfig(documentURL, out webConfigPath);
	
			// Add httphandler for mapcontrol
			bool updated = false;
			updated |= WebDesignerUtility.AddController(xmlDoc, "MapController.ashx", "MapInfo.WebControls.MapController");
			updated |= WebDesignerUtility.AddAppSettings(xmlDoc);
			updated |= WebDesignerUtility.AddHttpModules(xmlDoc);
			updated |= WebDesignerUtility.AddAssemblies(xmlDoc);
			updated |= WebDesignerUtility.AddSessionState(xmlDoc);

			if (updated) xmlDoc.Save(webConfigPath);
		}

		/// <summary>
		/// Initializes the designer.
		/// </summary>
		/// <param name="component">Component for which this designer is for.</param>
		/// <remarks>It sets up all parameters for the designer and updates web.config file.</remarks>
		public override void Initialize(IComponent component) 
		{
			_mapControl = component as MapControl;
			IDesignerHost host = (IDesignerHost)component.Site.GetService(typeof(IDesignerHost));
			IWebFormsDocumentService docService = (IWebFormsDocumentService)component.Site.GetService(typeof(IWebFormsDocumentService));
			string documentUrl = docService.DocumentUrl;
			UpdateForMapControl(documentUrl);
			base.Initialize(component);
		}

		/// <summary>
		/// Serialize the width and height of the mapcontrol
		/// </summary>
		/// <remarks>Serialize the width and height of the mapcontrol into the web form aspx file.</remarks>
		protected override void OnBehaviorAttached() {
			base.OnBehaviorAttached ();
			Behavior.SetAttribute("Width", _mapControl.Width.Value, false);
			Behavior.SetAttribute("Height", _mapControl.Height.Value, false);
		}

		/// <summary>
		/// Writes the design time HTML for MapControl. 
		/// </summary>
		/// <remarks>At design time no map is displayed, only a brief message indicating what it does at run time is visible 
		/// on the control.</remarks>
		/// <returns>string containing HTML</returns>
		public override string GetDesignTimeHtml() 
		{
			StringWriter text = new StringWriter();
			HtmlTextWriter writer = new HtmlTextWriter(text);
			writer.AddAttribute("style", string.Format("border-top:solid 1px black; border-left:solid 1px black; border-bottom:solid 1px black; border-right:solid 1px black; width: {0}px; height: {1}px;", _mapControl.Width.Value, _mapControl.Height.Value));
			writer.RenderBeginTag(HtmlTextWriterTag.Div);
			writer.WriteLine(L10NUtils.Resources.GetString("MCDesignString"));
			writer.RenderEndTag();
			return text.ToString();
		}
	}
}

File: MapXtreme/6.6/Samples/WebControlsSourceCode/vb/Design/MapControlDesigner.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.IO
Imports System.Web.UI.Design
Imports System.Web.UI
Imports System.Runtime.InteropServices
Imports EnvDTE
Imports System.Xml
Imports System.Reflection

Friend Class MapControlDesigner
    Inherits ControlDesigner
    Private _mapControl As MapControl

    'Updates the web.config file with information about MapXTreme assemblies and handlers.
    Public Shared Sub UpdateForMapControl(ByVal documentURL As String)
        ' Open the web.config file as xmldoc
        Dim webConfigPath As String
        Dim xmlDoc As XmlDocument = WebDesignerUtility.GetWebConfig(documentURL, webConfigPath)

        ' Add httphandler for mapcontrol
        Dim updated As Boolean = False
        updated = updated Or WebDesignerUtility.AddController(xmlDoc, "MapController.ashx", "MapInfo.WebControls.MapController")
        updated = updated Or WebDesignerUtility.AddAppSettings(xmlDoc)
        updated = updated Or WebDesignerUtility.AddHttpModules(xmlDoc)
        updated = updated Or WebDesignerUtility.AddAssemblies(xmlDoc)
        updated = updated Or WebDesignerUtility.AddSessionState(xmlDoc)

        If updated Then xmlDoc.Save(webConfigPath)
    End Sub

    Public Overrides Sub Initialize(ByVal component As IComponent)
        _mapControl = CType(component, MapControl)
        Dim host As IDesignerHost = CType(component.Site.GetService(GetType(IDesignerHost)), IDesignerHost)
        Dim docService As IWebFormsDocumentService = CType(component.Site.GetService(GetType(IWebFormsDocumentService)), IWebFormsDocumentService)
        Dim documentUrl As String = docService.DocumentUrl
        UpdateForMapControl(documentUrl)
        MyBase.Initialize(component)
    End Sub

    Protected Overrides Sub OnBehaviorAttached()
        MyBase.OnBehaviorAttached()
        Behavior.SetAttribute("Width", _mapControl.Width.Value, False)
        Behavior.SetAttribute("Height", _mapControl.Height.Value, False)
    End Sub

    Public Overrides Function GetDesignTimeHtml() As String
        Dim text As StringWriter = New StringWriter
        Dim writer As HtmlTextWriter = New HtmlTextWriter(text)
        writer.AddAttribute("style", String.Format("border-top:solid 1px black; border-left:solid 1px black; border-bottom:solid 1px black; border-right:solid 1px black; width: {0}px; height: {1}px;", _mapControl.Width.Value, _mapControl.Height.Value))
        writer.RenderBeginTag(HtmlTextWriterTag.Div)
        writer.WriteLine(L10NUtils.Resources.GetString("MCDesignString"))
        writer.RenderEndTag()
        Return text.ToString()
    End Function
End Class




